### PR TITLE
Remove chroma-diagnostics, replace with sosreport plugin

### LINKED ISSUE #263: Remove chroma-diagnostics, replace with sosreport plugin
We currently have dedicated code to collect diagnostic reports, `chroma-diagnostics`.

This code reinvents much of the wheel that is provided by `sosreport` + an `iml` plugin. 

Remove chroma-diagnostics, replace with a repo that hosts the `iml-sosreport-plugin`.

This repo can be published to pypi + added to copr and we can remove all `chroma-diagnostics` code.

### PR DESCRIPTION
**Pull request by jgrund**
Fixes #263

We currently have dedicated code to collect diagnostic reports, chroma-diagnostics.

This code reinvents much of the wheel that is provided by sosreport + an iml plugin.

Remove chroma-diagnostics, replace with a repo that hosts the iml-sosreport-plugin.

This repo can be published to pypi + added to copr and we can remove all chroma-diagnostics code.

---
This change is [Reviewable](https://reviewable.io/reviews/intel-hpdd/intel-manager-for-lustre/267)

### CODE REVIEW

File: chroma-manager/tests/integration/shared_storage_configuration/test_iml_diagnostics.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        run_result = self.remote_operations.run_iml_diagnostics(self.TEST_SERVER, verbose=True)
+        self.assertEqual(run_result.rc, 0,
+                         'IML Diagnostics failed with the following error: %s' % run_result.stderr)
+
+    def test_diagnostics_without_all_logs(self):
+
+        # Generate the diagnostics from the server
+        run_result = self.remote_operations.run_iml_diagnostics(self.TEST_SERVER, verbose=False)
+        self.assertEqual(run_result.rc, 0,
+                         'IML Diagnostics failed with the following error: %s' % run_result.stderr)
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/shared_storage_configuration/test_iml_diagnostics.py]: Should we expand these tests, or write new ones to (spot) test the content of the report?

File: chroma-manager/tests/integration/utils/chroma_log_collector.py [hunk 2]
```diff
@@ -135,46 +135,45 @@ def fetch_log_dir(self, server, dir):
...
-        # Check that chroma-diagnostics is installed. May not be if installation failed, etc.
-        if shell_run(['ssh', server, 'which chroma-diagnostics']).rc:
-            return["chroma-diagnostics not installed on %s. skipping." % server]
+        # Check that iml-diagnostics is installed. May not be if installation failed, etc.
+        if shell_run(['ssh', server, 'which iml-diagnostics']).rc:
+            return["iml-diagnostics not installed on %s. skipping." % server]
 
         # Generate the diagnostics from the server
-        result = shell_run(['ssh', server, 'chroma-diagnostics', '-v', '-v', '-v'])
+        result = shell_run(['ssh', server, 'iml-diagnostics', '--all-logs'])
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/utils/chroma_log_collector.py]: Do we really want `--all-logs` here?

File: chroma-manager/tests/integration/shared_storage_configuration/test_iml_diagnostics.py (hunk 1 above)
Comment (jgrund) [reply chroma-manager/tests/integration/shared_storage_configuration/test_iml_diagnostics.py]: Yes, but they should live in the iml-diagnostics repo when we do so (and I didn't realize we had these 😄  )

File: chroma-manager/tests/integration/utils/chroma_log_collector.py [hunk 3]
```diff
@@ -135,46 +135,45 @@ def fetch_log_dir(self, server, dir):
...
 
         if result.timeout:
```
[the remark this answers is not in the document]
Comment (jgrund) [reply chroma-manager/tests/integration/utils/chroma_log_collector.py]: `sos` does bound it's commands (not file copies).

[the remark this answers is not in the document]
Comment (jgrund) [reply chroma-manager/tests/integration/utils/chroma_log_collector.py]: I'll still update this though.

File: chroma-manager/tests/integration/utils/chroma_log_collector.py (hunk 2 above)
Comment (jgrund) [reply chroma-manager/tests/integration/utils/chroma_log_collector.py]: It's the closest equivalence to `-v -v -v` at the moment.

Given logs *may* roll over during tests (which I assume happens), I think we do.

File: chroma-manager/tests/integration/utils/chroma_log_collector.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply chroma-manager/tests/integration/utils/chroma_log_collector.py]: Great that SOS is time-bounding.

File: chroma-manager/tests/integration/utils/chroma_log_collector.py [hunk 4]
```diff
@@ -146,7 +146,8 @@ def fetch_iml_diagnostics(self, server):
             return["iml-diagnostics not installed on %s. skipping." % server]
 
         # Generate the diagnostics from the server
-        result = shell_run(['ssh', server, 'iml-diagnostics', '--all-logs'])
+        result = shell_run(
+            ['ssh', server, 'iml-diagnostics', '--all-logs'], timeout=300)
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/utils/chroma_log_collector.py]: I wonder if 5 minutes is enough.  I guess we can wait and see, but I would not be opposed to 10 minutes if you wanted.

I easy to wait and see if you prefer though.

Comment (jgrund) [reply chroma-manager/tests/integration/utils/chroma_log_collector.py]: Sure, I'll up it.

**Comment by jgrund** [comment]
Here is an example report I pulled during a test if interested:


[sosreport-lotus-44vm15-20170911120734.zip](https://github.com/intel-hpdd/intel-manager-for-lustre/files/1293801/sosreport-lotus-44vm15-20170911120734.zip)

**Comment by brianjmurrell** [comment]
> Here is an example report I pulled during a test if interested:
> 
> sosreport-lotus-44vm15-20170911120734.zip

Why does it have a `__MACOSX/` directory in it?

**Comment by jgrund** [comment]
> Why does it have a __MACOSX/ directory in it? 

Because I was lazy and repackaged the `.tar.xz` with OS X compress option. You can ignore those.

File: chroma-manager/tests/integration/utils/chroma_log_collector.py [hunk 5]
```diff
@@ -74,43 +81,58 @@ def shell_run(arg_list, timeout=0xfffffff):
...
+                               "%s-chroma_test.log" % test_runner))
+            errors.append(
+                self.fetch_log(test_runner, '/var/log/messages',
+                               "%s-messages.log" % test_runner))
 
         for server in self.chroma_managers + self.lustre_servers:
-            errors.append(self.fetch_log(server, '/var/log/yum.log', "%s-yum.log" % server))
-            errors.extend(self.fetch_chroma_diagnostics(server))
+            errors.append(
+                self.fetch_log(server, '/var/log/yum.log', "%s-yum.log" %
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/utils/chroma_log_collector.py]: Surely there is an SOS plugin to get this isn't there?

File: chroma-manager/tests/integration/utils/chroma_log_collector.py [hunk 6]
```diff
@@ -120,67 +142,97 @@ def fetch_log(self, server, source_log_path, destination_log_filename):
...
 
     def fetch_log_dir(self, server, dir):
-        """
-        Collect the log directory from the target
+        """Collect the log directory from the target.
+
         :return: None on success or error message that can be used for diagnosing what went wrong.
+
         """
         logs = shell_run(['ssh', server, "ls %s | xargs -n1 basename" % dir])
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/utils/chroma_log_collector.py]: Is this actually `logs = shell_run(['ssh', server, "ls %s | sed -e "s/.*\//" % dir])`?

File: chroma-manager/tests/integration/utils/chroma_log_collector.py [hunk 7]
```diff
@@ -120,67 +142,97 @@ def fetch_log(self, server, source_log_path, destination_log_filename):
...
         elif diagnostics.endswith('tar.gz'):
-            if shell_run(['tar', '-xvzf', "%s/%s" % (self.destination_path, diagnostics),
-                               '-C', self.destination_path]).rc:
-                errors.append("Error tar -xvzf the chroma diagnostics file")
+            if shell_run([
+                    'tar', '-xvzf',
+                    "%s/%s" % (self.destination_path, diagnostics), '-C',
+                    self.destination_path
+            ]).rc:
+                errors.append("Error tar -xvzf the iml diagnostics file")
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/utils/chroma_log_collector.py]: Would be nice to DRY out the above code by simply setting the tar decompresison switch in the if and applying that to a single `shell_run()` with something like:

```
if diagnostics.endswith('tar.xz'):
￼  sw = "J"
elif diagnostics.endswith('tar.gz'):
￼  sw = "z"
...
if shell_run([
￼        'tar', '-xv%sf',
        "%s/%s" % (sw, self.destination_path, diagnostics), '-C',
        self.destination_path
]).rc:
    errors.append("Error tar -xvzf the iml diagnostics file")
```

File: chroma-manager/tests/integration/utils/chroma_log_collector.py [hunk 8]
```diff
@@ -120,67 +142,97 @@ def fetch_log(self, server, source_log_path, destination_log_filename):
...
+            errors = "Didn't recognize iml-diagnostics file format"
+
+        diagnostics_dir = re.compile('(sosreport-.*)\.tar\..*').search(
+            diagnostics).group(1)
+
+        try:
+            os.chmod(
+                "%s/%s/sys/module/md_mod/parameters/new_array"
+                % (self.destination_path, diagnostics_dir),
+                0777
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/utils/chroma_log_collector.py]: So this is the only file that was choking the `chmod -R 777`?  What is it's permission flags before the `chmod()`?

File: chroma-manager/tests/integration/utils/chroma_log_collector.py [hunk 9]
```diff
@@ -120,67 +142,97 @@ def fetch_log(self, server, source_log_path, destination_log_filename):
...
+            )
+        except OSError:
+            pass
+
+        if shell_run(
+            ['rm', '-f',
+             "%s/%s" % (self.destination_path, diagnostics)]).rc:
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/utils/chroma_log_collector.py]: `shutil.rmtree()`?

File: chroma-manager/tests/integration/utils/chroma_log_collector.py [hunk 10]
```diff
@@ -120,67 +142,97 @@ def fetch_log(self, server, source_log_path, destination_log_filename):
...
-        if shell_run(['ssh', server, 'which chroma-diagnostics']).rc:
-            return["chroma-diagnostics not installed on %s. skipping." % server]
+        # Check that iml-diagnostics is installed. May not be if installation failed, etc.
+        if shell_run(['ssh', server, 'which iml-diagnostics']).rc:
+            return ["iml-diagnostics not installed on %s. skipping." % server]
 
         # Generate the diagnostics from the server
-        result = shell_run(['ssh', server, 'chroma-diagnostics', '-v', '-v', '-v'])
+        result = shell_run(
+            ['ssh', server, 'iml-diagnostics', '--all-logs'], timeout=600)
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/utils/chroma_log_collector.py]: Should[n't] we re-use our built-in _ssh_ based on Paramiko for this?

File: chroma-manager/tests/integration/utils/chroma_log_collector.py [hunk 11]
```diff
@@ -120,67 +142,97 @@ def fetch_log(self, server, source_log_path, destination_log_filename):
...
+        :return: empty list on success or list of error messages that can be
+            used for diagnosing what went wrong.
+
         """
 
-        # Check that chroma-diagnostics is installed. May not be if installation failed, etc.
-        if shell_run(['ssh', server, 'which chroma-diagnostics']).rc:
-            return["chroma-diagnostics not installed on %s. skipping." % server]
+        # Check that iml-diagnostics is installed. May not be if installation failed, etc.
+        if shell_run(['ssh', server, 'which iml-diagnostics']).rc:
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/utils/chroma_log_collector.py]: Should[n't] we re-use our built-in ssh based on Paramiko for this?

File: chroma-manager/tests/integration/utils/chroma_log_collector.py [hunk 12]
```diff
@@ -120,67 +142,97 @@ def fetch_log(self, server, source_log_path, destination_log_filename):
...
         logs = shell_run(['ssh', server, "ls %s | xargs -n1 basename" % dir])
 
         if logs.rc:
             return "Failed fecthing log dir %s from %s" % (server, dir)
 
         for log in logs.stdout.split():
-            destination_log_filename = "%s-%s-%s" % (server, dir.strip('/').split('/')[-1], log)
-            self.fetch_log(server, "%s/%s" % (dir, log), destination_log_filename)
+            destination_log_filename = "%s-%s-%s" % (
+                server, dir.strip('/').split('/')[-1], log)
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/utils/chroma_log_collector.py]: I think `os.path.basename(dir)` is the portable (and cleaner looking) version of `dir.strip('/').split('/')[-1]`.

**Comment by jgrund** [comment]
@brianjmurrell given most of your comments are from existing code, and this would be valuable to have landed; would it be ok if I addressed them in a separate review?

**Comment by brianjmurrell** [comment]
@jgrund sure.

**Pull request merged**